### DISCUSSION TITLE
How to apply one contrast-stretch value across multiple images?

### DISCUSSION THREAD

**Discussion by memphis-10** [ImageMagick/ImageMagick#4025] [Digital Image Processing]
Hi,

Looking for the user forum and ended up here. I hope this is the right spot.

I'm trying to have IM contrast-stretch a series of images with the same RGB values it uses on the first image.  Is there a way to "keep" or "read" then re-apply contrast stretch values?

Tangent: if possible, I would like to take this one step further to get the initial RGB contrast-stretch values to use on other images at a later time.

Thanks so much!

**Comment by fmw42** [comment]
There is no current read-out of the clip values from contrast-stretch that I know about.  What is your exact command line? What is your IM versions and platform/OS? Other approaches are possible. For example contrast stretch the first image. Then use a process of matching mean/standard deviation (i.e. brightness and contrast). On Windows, user snibgo has a script to do that. I have one for Unix-like systems.

**Reply by fmw42** [reply]
You could also ask the developers for an enhancement to get the clip values from -contrast-stretch listed to the terminal as output

**Reply by memphis-10** [reply]
This is really helpful.  Thank you.  I'm working in both Windows (XP) and Mac Mojave.  Using an old 35mm scanner to capture negatives and inverting them.

I assumed it would be some trickery in the current state.  The command varies, depending on the image, but it is something like this for the whole folder on XP to convert from RAW to TIF: 

(*.RAW file)

rename %%f %%~nf.rgb
magick convert -endian LSB -size 3000x2000+16 -depth 16 -interlace plane %%~nf.rgb -compress none -interlace none -depth 16 -gamma 2.2 -contrast-stretch %%~nf.tif
del /q %%~nf.rgb

**Reply by fmw42** [reply]
P.S. On ImageMagick 7, use magick, not magick convert, for proper syntax.

**Reply by fmw42** [reply]
Is this a typo?

-size 3000x2000+16

What is +16 there?

**Comment by memphis-10** [comment]
Thanks for taking the time with me. The syntax was indeed left over from an old version on XP.

regarding the “+16”
This is the script that was shared in the scanner group (Pakon F-135).  It was explained at one point,  but I’d have to dig for the reason now.  It has something to do with the *.RAW starting file.

**Reply by fmw42** [reply]
It is not proper syntax. One uses WxH+X+Y for -resize, but not for -size. I do not know if ImageMagick will ignore the +16. I suspect it is a duplicate from the need to specify -density 16 and somehow the 16 got appended to the WxH. It is certainly not needed for RAW format, only the -density 16. See image format RGB at https://imagemagick.org/script/formats.php#supported where it says:


_RGB | RW | Raw red, green, and blue samples | Use -size and -depth to specify the image width, height, and depth. To specify a single precision floating-point format, use -define quantum:format=floating-point. Set the depth to 32 for single precision floats, 64 for double precision, and 16 for half-precision._



Also I do not know Windows scripting, but it looks like you have no arguments for -contrast-stretch (which it should) and %%~nf.tif is all part of the output file name. Is that a correct reading or is %% the argument for -contrast-stretch which should have a space between it and the output file. Can you provide your command with actual names and arguments rather than variables?

**Comment by snibgo** [comment]
On "-size 3000x2000+16": see https://imagemagick.org/script/command-line-options.php#size :
```
-size width[xheight][+offset]
```
"+16" will ignore the first 16 bytes in the file, assuming them to be a header before pixel data starts.

**Reply by fmw42** [reply]
Thanks, snibgo. Learned something new!

**Comment by snibgo** [comment]
On "-contrast-stretch", see https://imagemagick.org/script/command-line-options.php#contrast-stretch . It needs an argument. When I try it with no argument, I get:
```
f:\web\im>%IMG7%magick -size 1x10 gradient:#123-#abc -contrast-stretch txt:

magick: MissingArgument `-contrast-stretch' at CLI arg 4 @ fatal/magick-cli.c/ProcessCommandOptions/447.
```

**Reply by memphis-10** [reply]
Thank you @snibgo!  I knew there was a reason and was about to dig back through the scanner user group to get the answer.  Yes, the scanner saves linear *.RAW images with a header.

And I appreciate you both catching the mistake with -contrast-stretch arguments.  I'm pulled in a million different directions this week and copied the wrong one.  In answer to your question @fmw42, _%%~nf.tif_ is, indeed, part of the file rename - a necessary step for the entire image processing.  

This should be the correct conversion from *.RAW scanner files to processed *.TIF:
- rename *.RAW to *.RGB
- *.RGB magick -endian LSB -size 3000x2000+16 -depth 16 -interlace plane -compress none -interlace none -depth 16 -gamma 2.2 -contrast-stretch 0.15x0.05%
- rename *.RGB to *.TIF

The Windows script I copied above in my first reply works for a whole folder (when -contrast-stretch arguments are present).  Aside from mistakes I may have made converting that script to a readable string here, I'm still hoping to get the -contrast-stretch values from one image to reuse on another.  Or another approach to achieving the same thing.  I tried opening an "issue" but it was moved to a discussion (which makes sense).  I see the -contrast-stretch operator is under review for future development but I'm new here and not sure if my input to the developers is welcome or warranted.  I know they already have a lot on their respective plates.

**Reply by fmw42** [reply]
The input is always welcome. But there is no guaranteed time for implementation. It depends upon how hard it would be and time available and priorities.

**Reply by fmw42** [reply]
As I mentioned before, you can process one image and then try to match mean and standard deviation on all the rest to the first one. Snibgo has a script to do that on Windows and I have one on Unix-like systems. I do not know how well that would work. But until you get the change to -contrast-stretch, you can try that. As you are on Windows, I cannot help with developing a script to get you the information you need. But it would not be too hard. The histogram simply needs to be cumulated and the gray level values determined from the bins where the percent are reached along the cumulated histogram from the ends. This could be scripted. But certainly it would be better if the -contrast-stretch code was simply modified to get the clip gray level values.

**Comment by snibgo** [comment]
It would be useful if "-contrast-stretch" could output metadata so we could then use a "-level" that gave the same effect.

I use two workarounds.

1. I have a "-process" module that creates a cumulative histogram. By cumulating that, I have a direct readout of what value to use as black-point (or white-point) to get a certain percentage of black (or white) clipping.

2. Suppose you have the input and output to a "-contrast-stretch". Select two pixels from the output that have different values that have not clipped. Select the corresponding input pixels. A calculation then gives the "-level" that would have the same effect.

As fmw42 says, you can also do this by finding the mean and standard-deviation of the input and output. However, "-contrast-stretch" causes clipping, even when HDRI is used, and this reduces accuracy.

**Reply by memphis-10** [reply]
Thank you both!  It will be about 2 weeks before I'm back on this project, but I will revisit then when I can focus on the suggestions you've provided.  Thanks a ton for your time and input!

**Reply by memphis-10** [reply]
@snibgo, I'm back on the project and wanted to follow up.  @fmw42 Mentioned a Windows script of yours that may be useful?  My scanner runs on XP, so I have that to work with.  

For simplicity, as I am a noob-novice at scripting from scratch (though I can read and adapt written ones pretty well), what would you suggest as a good starting point?  

For clarity: I am using contrast stretch with very minimal clipping to remove the base color from scanned negatives.  I understand any clipping on the initial -contrast-stretched image would affect the final outcome, but the likelihood of sampling 1 or 2 clipped pixels out of 6MP is low enough to try, and I can adapt accordingly if that happens.  Ultimately, I'd go through trial and error to obtain a set of -level args to use for each film base, so contrast stretch is only the foundation to discover my -level args for repeated use.

thanks a ton

**Reply by memphis-10** [reply]
>     2. Suppose you have the input and output to a "-contrast-stretch". Select two pixels from the output that have different values that have not clipped. Select the corresponding input pixels. A calculation then gives the "-level" that would have the same effect.

Hi @snibgo.  I'm back to this project after a nearly full spin around the sun.  The above method seems best for me at this moment.  I have the pixel values, but cannot figure out the calculation.  I'm sure it's staring right at me.  Do you mind offering some input?

stretched:
pixel 1: 162,144,144
pixel 2: 168,134,119

straight:
pixel 1: 74,63,59
pixel 2: 69,57,53 

Thank you!

**Reply by snibgo** [reply]
You have given pixel values, but have not said what the maximum value is. I assume it is 255. I prefer to work in percentages, so the maximum is 100. We convert by multiplying by 100/255.

stretched:
y1 = 63.53 56.47 56.47
y2 = 65.88 52.55 46.67

straight:
x1 = 29.02 24.71 23.14
x2 = 27.06 22.35 20.78

We have 3 channels: red, green and blue. In each channel, we have two input values (call these "x1" and "x2") and two output values (call these "y1" and "y2"). Imagine a graph with x and y axes. For two values of x, we know y. These two points define a straight line.

We assume a linear transformation from x to y: y = a*x+b, and then calculate a and b. We have simultaneous equations:
```
y1 = a * x1 + b
y2 = a * x2 + b
```
By subtraction:
```
y1 - y2 = a * (x1 - x2)
```
So:
```
a = (y1 - y2) / (x1 - x2)
```
So we can calculate a. Use this in either of the two simultaneous equations to calculate b.
```
b = y1 - a * x1
b = y2 - a * x2
```

"a" is the slope of the transformation. "b" is the intercept (the y value) at x=0. To get the intercept at x=100, we need to add the slope multiplied by 100.
```
v2 = b + a * 100
```
The values "b" and "v2" are exactly the percentages we need for the "+level" operation. (Note: "+" not "-". See https://imagemagick.org/script/command-line-options.php#level )

For example, consider the red channel.
```
a = (63.53 - 65.88) / (29.02 - 27.06) = -1.198979591836735
b = 63.53 - (-1.198979591836735 * 29.02) = 98.32438775510205
v2 = 98.32438775510205 + (-1.198979591836735 * 100) = -21.57357142857145
```
So the operation for the red channel is:
```
+level 98.324,-21.574%
```
We can check that the result is reasonable like this:
```
magick -size 1x10 gradient:Black-White +level 98.324,-21.574% txt:

# ImageMagick pixel enumeration: 1,10,0,65535,srgb
0,0: (64437,64437,64437)  #FBB5FBB5FBB5  srgb(98.324%,98.324%,98.324%)
0,1: (55706,55706,55706)  #D99AD99AD99A  srgb(85.002%,85.002%,85.002%)
0,2: (46975,46975,46975)  #B77FB77FB77F  srgb(71.68%,71.68%,71.68%)
0,3: (38245,38245,38245)  #956595659565  srgb(58.358%,58.358%,58.358%)
0,4: (29514,29514,29514)  #734A734A734A  srgb(45.036%,45.036%,45.036%)
0,5: (20784,20784,20784)  #513051305130  srgb(31.714%,31.714%,31.714%)
0,6: (12053,12053,12053)  #2F152F152F15  srgb(18.392%,18.392%,18.392%)
0,7: (3323,3323,3323)  #0CFB0CFB0CFB  srgb(5.07%,5.07%,5.07%)
0,8: (0,0,0)  #000000000000  srgb(-8.252%,-8.252%,-8.252%)
0,9: (0,0,0)  #000000000000  srgb(-21.574%,-21.574%,-21.574%)
```
Repeat the calculation for the green and blue channels.

**Reply by memphis-10** [reply]
Cheers, thank you for the formulas - and especially the explanation.  My mom was a math teacher and I still would not have figured this out any time soon.

I'm trying to implement the work and keep getting unexpected results.  I've triple checked my work and compared to your example, and the manual levels are still not a close match to the stretched image.

Supposing my output selections and calculations are all correct, is the the correct usage?
`-channel r +level -44.89,329.55% -channel g +level -53.34,439.13% -channel b +level -38.18,372.05%`

I chose a different sample since I believe there may have been a mistake with the first one.  Here's what I have (in percentages):
```
Stretched:
y1: 50.7 28.9 39.35
y2: 20.22 24.32 17.41

Straight:
x1: 25.53 16.7 18.9
x2: 17.39 15.77 13.55
```

for red
```
a = (50.7 - 20.22) / (25.53 - 17.39) = 3.74
b = 50.7 - (3.74 * 25.53) = -44.9
v2 = -44.9 + (3.74 * 100) = 329.55
+level -44.9,329.55%
```
thanks again for your time

**Reply by snibgo** [reply]
Your post is entirely correct. After doing the channel calculations, include "+channel". This ensures that any other operations will be done on all channels. For example (Windows CMD syntax):
```
magick ^
  in.tiff ^
  -channel r +level -44.89,329.55% ^
  -channel g +level -53.34,439.13% ^
  -channel b +level -38.18,372.05% ^
  +channel ^
  out.tiff
```
If this is in a BAT script, double the percent signs: "%%".

We can check the operations like this:
```
magick xc:gray(25.53%) +level -44.9,329.55% txt:

# ImageMagick pixel enumeration: 1,1,0,65535,gray
0,0: (33224)  #81C881C881C8  gray(50.6971%)


magick xc:gray(17.39%) +level -44.9,329.55% txt:

# ImageMagick pixel enumeration: 1,1,0,65535,gray
0,0: (13249)  #33C133C133C1  gray(20.2169%)
```
So the red channel transforms 25.53% to 50.6971%, and 17.39% to 20.2169%, as required.

**Reply by memphis-10** [reply]
Viola.  Thank you!  Not sure where my error was; may have simply been time of day and brain fatigue.  This is huge; thank you.  I'm working on some images with noise, so the minimal clipping is acceptable.

One more question, is there math to translate the new values for use in -level-colors "rgb(x,x,x,)","rgb(x,x,x)"?

I'm assuming I need to know the current "zero" values?  If that's right, I know one way I can get those is from a Gimp csv of the histogram.

**Reply by memphis-10** [reply]
I came back to add: I discovered a way to get RGB -level-colors values using the CSV.

Add the pixels on the top and bottom end to the allowable % of clipping, then use the RGB value where that total lies.  Where the total % falls between 2 values, get the fraction of the next (or previous) value that brings the total to the correct #/%

I'm happy to have made some progress here, and feel like this was staring at me for months.  I'm still interested in knowing if the original method above can be used to calculate the same values?

**Reply by snibgo** [reply]
> One more question, is there math to translate the new values for use in -level-colors "rgb(x,x,x,)","rgb(x,x,x)"?

No translation is required. The following two commands do the same thing (Windows BAT syntax):
```
magick ^
  %INFILE% ^
  -channel r +level -44.89,329.55%% ^
  -channel g +level -53.34,439.13%% ^
  -channel b +level -38.18,372.05%% ^
  +channel ^
  x0.png

magick ^
  %INFILE% ^
  +level-colors sRGB(-44.89%%,-53.34%%,-38.18%%),sRGB(329.55%%,439.13%%,372.05%%) ^
  x1.png
```

> I'm assuming I need to know the current "zero" values?

Sorry, I don't understand. What "zero" values?

> If that's right, I know one way I can get those is from a Gimp csv of the histogram.

Yes. As I said upthread, if we have a cumulative histogram, we can directly use that to find the cutoffs for that give the desired clipping. That is how "-contrast-stretch" works internally.

>  I'm still interested in knowing if the original method above can be used to calculate the same values?

Again, sorry, I don't understand the question.

**Reply by memphis-10** [reply]
> No translation is required. The following two commands do the same thing (Windows BAT syntax):
> ```
> magick ^
>   %INFILE% ^
>   -channel r +level -44.89,329.55%% ^
>   -channel g +level -53.34,439.13%% ^
>   -channel b +level -38.18,372.05%% ^
>   +channel ^
>   x0.png
> 
> magick ^
>   %INFILE% ^
>   +level-colors sRGB(-44.89%%,-53.34%%,-38.18%%),sRGB(329.55%%,439.13%%,372.05%%) ^
>   x1.png
> ```
Yes, I got it! Thank you!  This is great, I'm just curious about converting these % values into RGB numbers to set levels.
> > I'm assuming I need to know the current "zero" values?
> 
> Sorry, I don't understand. What "zero" values?

Ignore this, I'm not speaking in clear terms.

> > I'm still interested in knowing if the original method above can be used to calculate the same values?
> 
> Again, sorry, I don't understand the question.

Sorry for the confusion.  Seeing this used in +level-colors is hugely helpful.  In fitting it into my workflow, I was curious if it's possible to convert the % values you helped me get into -level-colors RGB numbers.

Thanks again

**Reply by snibgo** [reply]
"-level" and "-level-colors" are equivalent in the same way as "+level" and "+level-colors".

To use "-level", the math needs to change. For "+level", we found the y values at x=0 and x=100%. If you want to use "-level" instead, then you need to find the x values at y=0 and y=100%.

If you want a scale of 0-255 instead of 0-100, the process is identical. Where I have written "100" or "100%", you would have "255".

**Reply by memphis-10** [reply]
Thank you.  I'm beginning to get the hang of it and have created a calculator for quicker calculations of the first example.

One thing I didn't catch initially: the method you shared with me ends in a negative.  I missed this at first since I was working with negatives and my string includes "-negate" so I was used to seeing a positive at the end.  

Any help would be appreciated.  Thank you.

**Reply by snibgo** [reply]
> ... the method you shared with me ends in a negative. I missed this at first since I was working with negatives and my string includes "-negate" so I was used to seeing a positive at the end.

Sorry, I don't understand. The math I showed has many negatives. What is the problem?

"-negate" in a magick command is different.

**Reply by memphis-10** [reply]
Sorry, I meant the image was returning negative, but I think I found my mistake.  Still working on it.  Thanks again!

**Comment by snibgo** [comment]
> ... a Windows script of yours that may be useful?

I don't know which of my scripts might be helpful. My main page is [Snibgo's ImageMagick pages](http://im.snibgo.com/index.htm)

I have not digitized colour negatives. But I don't think that deliberately clipping data is a good idea.

The orange mask is a simple multiplier for two out of the three linear RGB channels. This is the same as a white balance.

**Reply by memphis-10** [reply]
White Balance is a great option, too.  thanks for the link--you have a killer collection there.  I appreciate the help!
